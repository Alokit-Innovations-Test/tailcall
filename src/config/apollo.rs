use anyhow::Context;
use serde::{Deserialize, Serialize};

use crate::config::ConfigReaderContext;
use crate::mustache::Mustache;

#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Apollo {
    ///
    /// Setting `apiKey` for Apollo.
    pub api_key: String,
    ///
    /// Setting `graphRef` for Apollo in the format <graphId>@<variant>.
    pub graph_ref: String,
    ///
    /// Setting `userVersion` for Apollo.
    #[serde(default = "default_user_version")]
    pub user_version: String,
    ///
    /// Setting `platform` for Apollo.
    #[serde(default = "default_platform")]
    pub platform: String,
    ///
    /// Setting `version` for Apollo.
    #[serde(default = "default_version")]
    pub version: String,
}

fn default_user_version() -> String {
    "1.0".to_string()
}

fn default_platform() -> String {
    "platform".to_string()
}

fn default_version() -> String {
    "1.0".to_string()
}

impl Apollo {
    pub fn render_mustache(&mut self, reader_ctx: &ConfigReaderContext) -> anyhow::Result<()> {
        let Apollo { api_key, graph_ref, user_version, platform, version } = self;

        let api_key_tmpl = Mustache::parse(api_key)?;
        *api_key = api_key_tmpl
            .render_string(reader_ctx)
            .context("apiKey is not defined")?;

        let graph_ref_tmpl = Mustache::parse(graph_ref)?;
        *graph_ref = graph_ref_tmpl
            .render_string(reader_ctx)
            .context("graphRef is not defined")?;

        let user_version_tmpl = Mustache::parse(user_version)?;
        if let Some(rendered) = user_version_tmpl.render_string(reader_ctx) {
            *user_version = rendered;
        }

        let platform_tmpl = Mustache::parse(platform)?;
        if let Some(rendered) = platform_tmpl.render_string(reader_ctx) {
            *platform = rendered;
        }

        let version_tmpl = Mustache::parse(version)?;
        if let Some(rendered) = version_tmpl.render_string(reader_ctx) {
            *version = rendered;
        }

        Ok(())
    }
}
