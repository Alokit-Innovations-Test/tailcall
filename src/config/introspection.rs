use std::collections::BTreeMap;

use async_graphql::futures_util::future::join_all;
use reqwest;
use serde_json::json;

// GraphQL introspection response types.
#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionResult {
  pub data: Option<IntrospectionData>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionData {
  pub __schema: IntrospectionSchema,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionSchema {
  #[serde(rename = "queryType")]
  pub query_type: IntrospectionType,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionType {
  pub fields: Vec<Field>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Field {
  pub name: String,
  pub args: Option<Vec<Arg>>,
  #[serde(rename = "type")]
  pub type_: Option<Type_>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Arg {
  pub name: String,
  #[serde(rename = "type")]
  pub type_: Type_,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Type_ {
  pub name: Option<String>,
  pub kind: Option<String>,
  #[serde(rename = "ofType")]
  pub of_type: Option<Box<Type_>>,
  pub fields: Option<Vec<Field>>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum TypeKind {
  Scalar,
  Object,
  Interface,
  Union,
  Enum,
  InputObject,
  List,
  NonNull,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Default)]
pub struct IntrospectionResults(pub BTreeMap<String, IntrospectionResult>);

impl IntrospectionResults {
  pub fn merge_right(mut self, other: Self) -> Self {
    let mut merged = self.0;
    for (key, value) in other.0 {
      merged.insert(key, value);
    }
    self.0 = merged;
    self
  }
}

pub async fn introspect_endpoint(graphql_url: &String) -> Result<IntrospectionResult, Box<dyn std::error::Error>> {
  let introspection_query: String =
  json!({"query": "query { __schema { queryType { name fields { name args { name type { name kind ofType {name }}} type { name kind ofType { name fields {name} } fields { name }}  } } } }"}).to_string();

  let result = reqwest::Client::new()
    .post(graphql_url)
    .header("Content-Type", "application/json")
    .body(introspection_query)
    .send()
    .await?
    .json::<IntrospectionResult>()
    .await?;

  Ok(result)
}

pub async fn introspect_endpoints(graphql_urls: Vec<String>) -> IntrospectionResults {
  let mut results = BTreeMap::new();
  let introspect_futures = graphql_urls
    .iter()
    .map(|url| async move { (url.clone(), introspect_endpoint(url).await.unwrap()) });
  let joined = join_all(introspect_futures).await;
  for (url, introspection_result) in joined {
    results.insert(url.clone(), introspection_result.clone());
  }
  IntrospectionResults(results)
}

pub fn get_arg_type(
  introspection_result: &IntrospectionResult,
  query_name: &String,
  arg_name: &String,
) -> Option<String> {
  introspection_result
    .data
    .as_ref()
    .and_then(|data| {
      data
        .__schema
        .query_type
        .fields
        .iter()
        .find(|field| field.name == *query_name)
    })
    .and_then(|field| field.args.as_ref())
    .and_then(|args| args.iter().find(|arg| arg.name == *arg_name))
    .and_then(|arg| {
      arg.type_.name.as_ref().map_or_else(
        || {
          let kind = arg.type_.kind.clone().unwrap_or_default();
          match &arg.type_.of_type {
            Some(type_) => type_.name.as_ref().map(|name| format_name(name, &kind)),
            None => None,
          }
        },
        |name| Some(name.clone()),
      )
    })
}

fn format_name(name: &String, kind: &String) -> String {
  if kind == "LIST" {
    format!("[{}]", name)
  } else if kind == "NON_NULL" {
    format!("{}!", name)
  } else {
    name.clone()
  }
}

#[cfg(test)]
mod tests {
  use std::fs;
  use std::path::PathBuf;

  use httpmock::Method::POST;
  use httpmock::MockServer;

  use crate::config::introspection::*;

  fn load_mock_introspection_result() -> String {
    let mut mock_result_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    mock_result_path.push("tests/data/introspection-result.json");
    fs::read_to_string(mock_result_path).unwrap()
  }

  fn mock_introspection_results() -> BTreeMap<String, IntrospectionResult> {
    let result: IntrospectionResult = serde_json::from_str(&load_mock_introspection_result()).unwrap();
    let mut introspection_result: BTreeMap<String, IntrospectionResult> = BTreeMap::new();
    introspection_result.insert("http://localhost:8000/graphql".to_string(), result);
    introspection_result
  }

  #[test]
  fn test_get_arg_type() {
    let introspection_results = mock_introspection_results();
    let introspection_result = introspection_results
      .get(&"http://localhost:8000/graphql".to_string())
      .unwrap();
    assert_eq!(
      get_arg_type(introspection_result, &"user".to_string(), &"id".to_string()).unwrap(),
      "Int"
    );
    assert_eq!(
      get_arg_type(introspection_result, &"post".to_string(), &"id".to_string()).unwrap(),
      "Int!"
    );
    assert_eq!(
      get_arg_type(introspection_result, &"post".to_string(), &"doesntexist".to_string()),
      None
    );
    assert_eq!(
      get_arg_type(introspection_result, &"doesntexist".to_string(), &"id".to_string()),
      None
    );
  }

  #[tokio::test]
  async fn test_introspect_endpoint() {
    let server = MockServer::start();
    let result = load_mock_introspection_result();

    server.mock(|when, then| {
      when.method(POST).path("/graphql");
      then.status(200).header("content-type", "application/json").body(result);
    });

    let graphql_url = server.url("/graphql");

    let introspect_result = introspect_endpoint(&graphql_url).await;
    assert_eq!(
      introspect_result
        .unwrap()
        .data
        .unwrap()
        .__schema
        .query_type
        .fields
        .len(),
      2
    );
  }
}
