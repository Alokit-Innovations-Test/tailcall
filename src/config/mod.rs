mod config;
mod from_document;
pub mod group_by;
mod into_document;
mod key_values;
mod n_plus_one;
mod server;
mod source;
mod js_plugin;

pub use config::*;
pub use key_values::*;
pub use server::*;
pub use source::*;
pub use js_plugin::*;

fn is_default<T: Default + Eq>(val: &T) -> bool {
  *val == T::default()
}
