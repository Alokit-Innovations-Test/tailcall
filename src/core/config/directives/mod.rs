use lazy_static::lazy_static;

mod add_field;
mod alias;
mod cache;
mod call;
mod expr;
mod federation;
mod graphql;
mod grpc;
mod http;
mod js;
mod link;
mod modify;
mod omit;
mod protected;
mod server;
mod telemetry;
mod upstream;

pub use add_field::*;
pub use alias::*;
pub use cache::*;
pub use call::*;
pub use expr::*;
pub use federation::*;
pub use graphql::*;
pub use grpc::*;
pub use http::*;
pub use js::*;
pub use link::*;
pub use modify::*;
pub use omit::*;
pub use protected::*;
pub use server::*;
pub use telemetry::*;
pub use upstream::*;

use crate::core::directive::DirectiveCodec;

lazy_static! {
    pub static ref KNOWN_DIRECTIVES: Vec<String> = vec![
        add_field::AddField::directive_name(),
        alias::Alias::directive_name(),
        cache::Cache::directive_name(),
        call::Call::directive_name(),
        expr::Expr::directive_name(),
        graphql::GraphQL::directive_name(),
        grpc::Grpc::directive_name(),
        http::Http::directive_name(),
        js::JS::directive_name(),
        link::Link::directive_name(),
        modify::Modify::directive_name(),
        omit::Omit::directive_name(),
        protected::Protected::directive_name(),
        server::Server::directive_name(),
        telemetry::Telemetry::directive_name(),
        upstream::Upstream::directive_name(),
    ];
}
