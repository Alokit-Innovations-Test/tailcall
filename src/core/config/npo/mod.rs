mod identifier;
mod queries;

use std::{collections::HashSet, rc::Rc};

pub use identifier::*;
pub use queries::*;

use super::Config;

struct NPOIdentifier<'a> {
    config: &'a Config,
}

impl<'a> NPOIdentifier<'a> {
    pub fn new(config: &'a Config) -> NPOIdentifier {
        NPOIdentifier { config }
    }

    fn iter(
        &self,
        path: Chunk<FieldName<'a>>,
        type_name: TypeName<'a>,
        is_list: bool,
        visited: &mut HashSet<TypeName<'a>>,
    ) -> Chunk<Chunk<FieldName<'a>>> {
        if visited.contains(&type_name) {
            return Chunk::new();
        } else {
            visited.insert(type_name);
        }
        let mut chunks = Chunk::new();
        if let Some(type_of) = self.config.find_type(&type_name.as_str()) {
            for (name, field) in type_of.fields.iter() {
                let path = path.clone().append(FieldName::new(name));
                let is_batch = field.has_batched_resolver();

                if field.has_resolver() {
                    if !is_batch && is_list {
                        chunks = chunks.append(path.clone());
                    }
                }

                let is_list = is_list | field.list;

                chunks = chunks.concat(self.iter(
                    path,
                    TypeName::new(field.type_of.as_str()),
                    is_list,
                    visited,
                ))
            }
        }

        chunks
    }

    fn find_chunks(&self) -> Chunk<Chunk<FieldName<'a>>> {
        match &self.config.schema.query {
            None => Chunk::new(),
            Some(query) => self.iter(
                Chunk::new(),
                TypeName::new(query.as_str()),
                false,
                &mut HashSet::new(),
            ),
        }
    }

    fn find(&self) -> Vec<Vec<&'a str>> {
        self.find_chunks()
            .as_vec()
            .iter()
            .map(|chunk| {
                chunk
                    .as_vec()
                    .iter()
                    .map(|field_name| field_name.as_str())
                    .collect()
            })
            .collect()
    }
}

#[derive(Clone)]
enum Chunk<A> {
    Nil,
    Cons(A, Rc<Chunk<A>>),
    Concat(Rc<Chunk<A>>, Rc<Chunk<A>>),
}

impl<A> Chunk<A> {
    fn new() -> Self {
        Self::Nil
    }

    fn from_vec(vec: Vec<A>) -> Chunk<A> {
        let mut path = Chunk::new();
        for a in vec.into_iter().rev() {
            path = Chunk::Cons(a, Rc::new(path));
        }
        path
    }

    fn append(self, a: A) -> Self {
        Chunk::Cons(a, Rc::new(self))
    }

    fn concat(self, other: Chunk<A>) -> Self {
        Self::Concat(Rc::new(self), Rc::new(other))
    }

    fn as_vec(&self) -> Vec<&A> {
        let mut vec = Vec::new();
        self.as_vec_mut(&mut vec);
        vec
    }

    fn as_vec_mut<'a>(&'a self, buf: &mut Vec<&'a A>) {
        match self {
            Chunk::Nil => {}
            Chunk::Cons(a, rest) => {
                buf.push(a);
                rest.as_vec_mut(buf);
            }
            Chunk::Concat(a, b) => {
                a.as_vec_mut(buf);
                b.as_vec_mut(buf);
            }
        }
    }
}
