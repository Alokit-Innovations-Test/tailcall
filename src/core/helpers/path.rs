use std::fs;
use std::path::Path;

use pathdiff::diff_paths;

/// Checks if file or folder already exists or not.
pub fn is_exists(path: &str) -> bool {
    fs::metadata(path).is_ok()
}

/// Expects both paths to be absolute and returns a relative path from `from` to
/// `to`. expects `from`` to be directory.
pub fn to_relative_path(from: &Path, to: &str) -> Option<String> {
    let from_path = Path::new(from).to_path_buf();
    let to_path = Path::new(to).to_path_buf();

    // Calculate the relative path from `from_path` to `to_path`
    diff_paths(to_path, from_path).map(|p| p.to_string_lossy().to_string())
}
