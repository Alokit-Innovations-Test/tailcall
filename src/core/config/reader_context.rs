use std::borrow::Cow;
use std::collections::BTreeMap;

use headers::HeaderMap;
use serde::Deserialize;
use crate::core::has_headers::HasHeaders;
use crate::core::json::JsonLike;
use crate::core::path::PathString;
use crate::core::runtime::TargetRuntime;

pub struct ConfigReaderContext<'a, Value> {
    pub runtime: &'a TargetRuntime<Value>,
    pub vars: &'a BTreeMap<String, String>,
    pub headers: HeaderMap,
}

impl<'a, Value: JsonLike<'a> + Deserialize<'a> + Clone> PathString for ConfigReaderContext<'a, Value> {
    fn path_string<T: AsRef<str>>(&self, path: &[T]) -> Option<Cow<'_, str>> {
        if path.is_empty() {
            return None;
        }

        path.split_first()
            .and_then(|(head, tail)| match head.as_ref() {
                "vars" => self.vars.get(tail[0].as_ref()).map(|v| v.into()),
                "env" => self.runtime.env.get(tail[0].as_ref()),
                _ => None,
            })
    }
}

impl<Value> HasHeaders for ConfigReaderContext<'_, Value> {
    fn headers(&self) -> &HeaderMap {
        &self.headers
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use super::*;
    use crate::core::tests::TestEnvIO;

    #[test]
    fn path_string() {
        let mut runtime = crate::core::runtime::test::init(None);
        runtime.env = Arc::new(TestEnvIO::from_iter([(
            "ENV_1".to_owned(),
            "ENV_VAL".to_owned(),
        )]));

        let reader_context = ConfigReaderContext {
            runtime: &runtime,
            vars: &BTreeMap::from_iter([("VAR_1".to_owned(), "VAR_VAL".to_owned())]),
            headers: Default::default(),
        };

        assert_eq!(
            reader_context.path_string(&["env", "ENV_1"]),
            Some("ENV_VAL".into())
        );
        assert_eq!(reader_context.path_string(&["env", "ENV_5"]), None);
        assert_eq!(
            reader_context.path_string(&["vars", "VAR_1"]),
            Some("VAR_VAL".into())
        );
        assert_eq!(reader_context.path_string(&["vars", "VAR_6"]), None);
        assert_eq!(reader_context.path_string(&["unknown", "unknown"]), None);
    }
}
