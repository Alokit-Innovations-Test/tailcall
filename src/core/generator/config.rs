use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
