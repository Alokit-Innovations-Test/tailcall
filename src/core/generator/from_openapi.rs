use oas3::Spec;

use crate::core::config::Config;

#[derive(Default)]
pub struct OpenApiToConfigConverter {
    #[allow(unused)]
    spec: Spec,
    config: Config,
}

impl OpenApiToConfigConverter {
    pub fn new(query: impl AsRef<str>, spec_str: impl AsRef<str>) -> anyhow::Result<Self> {
        let spec = oas3::from_reader(spec_str.as_ref().as_bytes())?;
        let config = Config::default().query(query.as_ref());
        Ok(Self { config, spec })
    }

    pub fn convert(self) -> Config {
        self.config
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::Path;

    use super::*;

    #[test]
    fn test_openapi_apis_guru() {
        let apis_guru = config_from_openapi_spec("apis-guru.yml");
        insta::assert_snapshot!(apis_guru);
    }

    #[test]
    fn test_openapi_jsonplaceholder() {
        let jsonplaceholder = config_from_openapi_spec("jsonplaceholder.yml");
        insta::assert_snapshot!(jsonplaceholder);
    }

    #[test]
    fn test_openapi_spotify() {
        let spotify = config_from_openapi_spec("spotify.yml");
        insta::assert_snapshot!(spotify);
    }

    fn config_from_openapi_spec(filename: &str) -> String {
        let spec_path = Path::new("src")
            .join("core")
            .join("generator")
            .join("openapi")
            .join(filename);

        let content = fs::read_to_string(spec_path).unwrap();
        OpenApiToConfigConverter::new("Query", content.as_str())
            .unwrap()
            .convert()
            .to_sdl()
    }
}
