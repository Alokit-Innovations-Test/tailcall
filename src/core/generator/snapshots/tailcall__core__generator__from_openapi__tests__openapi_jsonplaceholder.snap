---
source: src/core/generator/from_openapi.rs
expression: jsonplaceholder
---
schema @server @upstream {
  query: Query
}
